File: Wayner Backend/app/schemas/promocion.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field, field_validator


class PromocionCreate(BaseModel):
    codigo_barra: str = Field(..., min_length=1, max_length=50)
    nombre_producto: str = Field(..., min_length=1, max_length=255)

    precio_base: Decimal = Field(default=Decimal("0.00"), ge=0)
    precio_anterior: Decimal = Field(..., gt=0)
    precio_actual_prom: Decimal = Field(..., ge=0)

    encabezado: str | None = Field(default=None, max_length=255)
    mecanica: str | None = Field(default=None, max_length=255)

    fecha_inicio: date
    fecha_fin: date
    activa: bool = True

    @field_validator("fecha_fin")
    @classmethod
    def validar_fechas(cls, value: date, info):
        fecha_inicio = info.data.get("fecha_inicio")
        if fecha_inicio and value < fecha_inicio:
            raise ValueError("La fecha fin no puede ser menor que la fecha inicio")
        return value


class PromocionUpdate(BaseModel):
    nombre_producto: str | None = Field(default=None, max_length=255)

    precio_base: Decimal | None = Field(default=None, ge=0)
    precio_anterior: Decimal | None = Field(default=None, gt=0)
    precio_actual_prom: Decimal | None = Field(default=None, ge=0)

    encabezado: str | None = Field(default=None, max_length=255)
    mecanica: str | None = Field(default=None, max_length=255)

    fecha_inicio: date | None = None
    fecha_fin: date | None = None
    activa: bool | None = None


class PromocionResponse(BaseModel):
    id: int
    codigo_barra: str
    nombre_producto: str
    precio_base: Decimal
    precio_anterior: Decimal
    precio_actual_prom: Decimal
    ahorro: Decimal
    encabezado: str | None
    mecanica: str | None
    fecha_inicio: date
    fecha_fin: date
    activa: bool
    created_at: datetime
    updated_at: datetime | None

File: Wayner App Flutter/lib/features/promociones/models/promocion.dart
class Promocion {
  final int id;
  final String codigoBarra;
  final String nombreProducto;
  final double precioBase;
  final double precioAnterior;
  final double precioActualProm;
  final double ahorro;
  final String? encabezado;
  final String? mecanica;
  final DateTime fechaInicio;
  final DateTime fechaFin;
  final bool activa;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Promocion({
    required this.id,
    required this.codigoBarra,
    required this.nombreProducto,
    required this.precioBase,
    required this.precioAnterior,
    required this.precioActualProm,
    required this.ahorro,
    required this.encabezado,
    required this.mecanica,
    required this.fechaInicio,
    required this.fechaFin,
    required this.activa,
    this.createdAt,
    this.updatedAt,
  });

  factory Promocion.fromJson(Map<String, dynamic> json) {
    double toDouble(dynamic value) {
      if (value == null) return 0;
      if (value is num) return value.toDouble();
      return double.tryParse(value.toString()) ?? 0;
    }

    return Promocion(
      id: json['id'] ?? 0,
      codigoBarra: json['codigo_barra']?.toString() ?? '',
      nombreProducto: json['nombre_producto']?.toString() ?? '',
      precioBase: toDouble(json['precio_base']),
      precioAnterior: toDouble(json['precio_anterior']),
      precioActualProm: toDouble(json['precio_actual_prom']),
      ahorro: toDouble(json['ahorro']),
      encabezado: json['encabezado']?.toString(),
      mecanica: json['mecanica']?.toString(),
      fechaInicio: DateTime.parse(json['fecha_inicio']),
      fechaFin: DateTime.parse(json['fecha_fin']),
      activa: json['activa'] == true,
      createdAt: json['created_at'] == null
          ? null
          : DateTime.tryParse(json['created_at'].toString()),
      updatedAt: json['updated_at'] == null
          ? null
          : DateTime.tryParse(json['updated_at'].toString()),
    );
  }
}